cli/src/plz/cli/configuration.py

```diff
@@ -83,41 +83,120 @@ class Configuration:
-        try:
-            file_configuration = Configuration.from_file(
-                Configuration.CONFIGURATION_FILE, Configuration.PROPERTIES)
-        except FileNotFoundError:
+    def load(configuration_path: Optional[str] = None) -> 'Configuration':
+        config_file_name = Configuration._configuration_file_from_path(
+            configuration_path)
+
+        configuration = Configuration.defaults(Configuration.PROPERTIES)
+        configuration, user_level_config_was_read = \
+            Configuration._override_with_user_level_config(configuration)
+
+        configuration, file_config_was_read = \
+            Configuration._override_with_file_configs(
+                config_file_name, configuration,
+                plz_config_set_explicitly=configuration_path is not None)
+
+        if not (user_level_config_was_read or file_config_was_read):
             raise ValidationException(
                 [Configuration.MISSING_CONFIGURATION_FILE_ERROR])
+
         env_configuration = Configuration.from_env(Configuration.PROPERTIES)
 
-        return default_configuration \
-            .override_with(file_configuration) \
-            .override_with(env_configuration) \
-            .validate()
+        return configuration.override_with(env_configuration).validate()
+
+    @staticmethod
+    def _override_with_file_configs(
+            config_file_name: str, configuration: 'Configuration',
+            plz_config_set_explicitly: bool) -> Tuple['Configuration', bool]:
+        config_was_read = False
+        # Load all plz.config.json in parent directories since the mount point
+        path_fragments = config_file_name.split(os.path.sep)[:-1]
+        mount_index = len(path_fragments) - 1
+        # Stopping at 0. The empty path is the first entry, if we consider
+        # path_fragments[:0] it'd be duplicated
+        for n in range(len(path_fragments), 0, -1):
+            mount_index = n
+            if os.path.ismount(os.path.join('/', *path_fragments[:n])):
+                break
+
+        for n in range(mount_index, len(path_fragments) + 1):
+            if n < len(path_fragments) - 1:
+                file_name = Configuration.DEFAULT_CONFIGURATION_FILE_NAME
+            else:
+                _, file_name = os.path.split(config_file_name)
+            file_configuration = Configuration.from_file(
+                os.path.join('/', *path_fragments[:n], file_name),
+                Configuration.PROPERTIES,
+                fail_on_read_error=config_was_read)
+            if file_configuration is not None:
+                config_was_read = True
+                configuration = configuration.override_with(file_configuration)
+            else:
+                if n == len(path_fragments) - 1 and plz_config_set_explicitly:
+                    raise ValidationException([
+                        ValidationError(
+                            f'Couldn\'t read from {config_file_name}')])
```

Comment:
Either break this apart, or add way more comments. I have no idea what all the different cases are for.